src/algebra/group/inj_surj.lean

```diff
@@ -109,6 +109,22 @@ protected def monoid [monoid M₂] (f : M₁ → M₂) (hf : injective f)
   monoid M₁ :=
 { .. hf.semigroup f mul, .. hf.mul_one_class f one mul }
 
+/-- A type endowed with `1` and `*` is a monoid,
+if it admits an injective map that preserves `1` and `*` to a monoid.
+This version takes a custom `npow` as a `[has_pow M₁ ℕ]` argument.
+See note [reducible non-instances]. -/
+@[reducible, to_additive add_monoid_smul
+"A type endowed with `0` and `+` is an additive monoid,
+if it admits an injective map that preserves `0` and `+` to an additive monoid.
+This version takes a custom `nsmul` as a `[has_scalar ℕ M₁]` argument."]
+protected def monoid_pow [has_pow M₁ ℕ] [monoid M₂] (f : M₁ → M₂) (hf : injective f)
```

Comment:
I would argue we should call this `monoid` and rename the existing `monoid` to `monoid_default_pow`. IMO, in the long run we do not really want to be using the default `npow` anywhere, meaning with my suggested name we eventually end up with the clearer `monoid` spelling everywere.

Reply:
While I agree that we should do this in the long run, I think that this renaming should be done together with migration of all/most applications to the new API. I would prefer to do this migration on a case by case basis, then search&replace `s/monoid_pow/monoid/`.

Reply:
In particular, I already have branches for `ulift` and `opposite`, and I would prefer to merge them before we rename definitions.

> `group` doesn't seem particularly special compared to these.

It is special because you can't define `ulift.group` as `{ .. ulift.div_inv_monoid, .. ulift.some_other_instance }`.

Reply:
Thanks, I get the idea behind `group` then.

I assume the same argument applies to `group_with_zero `?